### PR TITLE
feat: Add support for copying files in and out of stopped containers

### PR DESCRIPTION
**Pull request by vsiravar**
### What does this PR do 

This adds support to copy files in and out of non running containers **only in rootful mode** thereby  improving feature compatibility with podman and docker. It fixes https://github.com/containerd/nerdctl/issues/1058

### High level design

This PR adds logic in 4 packages

- ~~main(Modifies parsing logic for `nerdctl cp` command)~~
- types(Modifies `ContainerCpOptions` struct)
- container(Adds logic to search for container by long/short Id)
- containerutil(Adds logic to copy files out and into non running containers)
- rootlessutil(Add a function to retrieve rootless containerd socket address)

#### [cmd/nerdctl/container_cp_linux.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-a6cc11fee8c6a93b90ff0594ba0ad1550e9e95de1f7a5f544cd47b8751c82add)

- Modifies `processCpOptions` function to include `GlobalCommandOptions`. Deletes logic for `container inspect` as we don't rely on inspect to find `pid` of the running container. For rootless mode `options.GOptions.Address=/proc/<PID of containerd>/root/run/containerd/containerd.sock` based on [faq.md](https://github.com/containerd/nerdctl/blob/main/docs/faq.md#containerd-socket-address)
- ~~Modifies `cpAction` to include `containerd.Client` which is used for searching container by long/short id and by `SnapshotService` to get the underlying `Snapshotter`.~~

#### ~~[main_linux.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-f9cc0f02d389d208c96ceb05e4014d58dfebb5e8071b039836e7b1a055c04b6d)~~

- ~~Modifies `appNeedsRootlessParentMain` to return true for `nerdctl cp` and `nerdctl container cp`. This is necessary to access containerd.sock in rootless mode.~~

#### [pkg/api/types/container_types.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-ba3b04d57ead8362915cc38b5a4ad5162a5b6d49f47595097204639cbaacd6d2)

- Modifies `ContainerCpOptions` to add `GlobalCommandOptions` and `ContainerReq`. ` GlobalCommandOptions` are necessary for determining namespace, conatinerd socket address and name of underlying snapshotter. `ContainerReq` required for finding container by short/long id. 

#### [pkg/cmd/container/cp_linux.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-6dbbf4a38e8014b0d33feae0a7ea8107317000ae2d16a9f27ee47951f1f79f97)

- Searches for container specified in `nerdctl cp <containerid>` command using `ContainerWalker`. If multiple containers are found it returns an error. 
- It returns an error when the specified container for copy is not found. 

#### [pkg/containerutil/cp_linux.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-2167aef78f978368dfab2db8ef6002a9fbcf827896d649f0986744122ac3813d)

- Use [containerd snapshot service](https://github.com/containerd/containerd/blob/main/api/services/snapshots/v1/snapshots.proto#L29) to copy files into and out of created/stopped containers. 
- It tries to find `[Task](https://github.com/containerd/containerd/blob/main/task.go#L154)` from containerd. If it's not found(which is the case on `nerdctl container create`, it mounts the snapshot onto a temporary directory for copying. Any other type of error returns an error. 
- If the status of the containerd task is Running, then the root filesystem of the running process is used for copying files into /out of the container. (unchanged from previous commits)
- If task is found and status is not Running(Exited container), then we fallback to the snapshot service for copying. 
- ~~For rootless mode it deletes the nsenter command to enter the namespace of the container process since it's already in the namespace of the rootlesskit child pid.~~ 
- Introduces a helper function `mountSnapshotForContainer` which returns the temp directory that the container shapshot is mounted on. 

#### [pkg/rootlessutil/rootlessutil_linux.go](https://github.com/containerd/nerdctl/pull/2355/files#diff-116470a846123b134de3715c584f11eef78e330070ecf12bcfef9c478a0d0cfd)

- Adds a function to retrieve rootless containerd socket address

### Testing

Added tests for [stopped containers](https://github.com/containerd/nerdctl/pull/2355/files#diff-fc1f66854a3bb061fe38564a668b1b07c4be8fe9f404e8248b80ab2edad6e22f) for exactly the same existing test cases as running containers. Tests are skipped in rootless mode on stopped containers.

### CODE REVIEW

File: pkg/containerutil/cp_linux.go [hunk 1]
```diff
@@ -177,3 +201,31 @@ func CopyFiles(ctx context.Context, container2host bool, pid int, dst, src strin
...
+		return "", nil, err
+	}
+	tempDir, err := os.MkdirTemp("", "nerdctl-cp-")
+	if err != nil {
+		return "", nil, err
+	}
+	err = mount.All(resp, tempDir)
+	if err != nil {
+		cleanup := func() {
+			os.RemoveAll(tempDir)
```
Comment (ktock) [inline pkg/containerutil/cp_linux.go:222]: why not remove it now?

File: pkg/containerutil/cp_linux.go [hunk 2]
```diff
@@ -138,14 +170,6 @@ func CopyFiles(ctx context.Context, container2host bool, pid int, dst, src strin
...
 	}
 	tarX = append(tarX, "-f", "-")
-	if rootlessutil.IsRootless() {
-		nsenter := []string{"nsenter", "-t", strconv.Itoa(pid), "-U", "--preserve-credentials", "--"}
-		if container2host {
-			tarC = append(nsenter, tarC...)
-		} else {
-			tarX = append(nsenter, tarX...)
-		}
-	}
```
Comment (ktock) [inline pkg/containerutil/cp_linux.go:148]: Can it be safely removed when ` status.Status == containerd.Running`?

File: pkg/api/types/container_types.go [hunk 3]
```diff
@@ -408,9 +408,10 @@ type ContainerListOptions struct {
 
 // ContainerCpOptions specifies options for `nerdctl (container) cp`
 type ContainerCpOptions struct {
+	// GOptions is the global options.
+	GOptions       GlobalCommandOptions
+	ContainerReq   string
```
[the remark this answers is not in the document]
Comment (vsiravar) [reply pkg/api/types/container_types.go:414]: Updated.

File: pkg/containerutil/cp_linux.go [hunk 4]
```diff
@@ -177,3 +201,31 @@ func CopyFiles(ctx context.Context, container2host bool, pid int, dst, src strin
...
+	err = mount.All(resp, tempDir)
+	if err != nil {
+		cleanup := func() {
+			os.RemoveAll(tempDir)
+		}
+		return "", cleanup, err
+	}
+	cleanup := func() {
+		mount.Unmount(tempDir, 0)
+		os.RemoveAll(tempDir)
```
[the remark this answers is not in the document]
Comment (vsiravar) [reply pkg/containerutil/cp_linux.go]: Updated

File: pkg/containerutil/cp_linux.go (hunk 1 above)
Comment (vsiravar) [reply pkg/containerutil/cp_linux.go]: Updated

File: pkg/containerutil/cp_linux.go (hunk 2 above)
Comment (vsiravar) [reply pkg/containerutil/cp_linux.go:148]: I don't quite understand this part, in rootless mode its already in the [namespace of the rootlesskit child pid](https://github.com/containerd/nerdctl/blob/main/pkg/rootlessutil/parent_linux.go#L66). I am missing anything here? The integration tests pass okay when `status.Status == containerd.Running` is true.

**Comment by AkihiroSuda** [comment]
Please try rebasing to make the CI green

File: cmd/nerdctl/main_linux.go [hunk 5]
```diff
@@ -39,16 +39,15 @@ func appNeedsRootlessParentMain(cmd *cobra.Command, args []string) bool {
 	switch commands[1] {
 	// completion, login, logout: false, because it shouldn't require the daemon to be running
 	// apparmor: false, because it requires the initial mount namespace to access /sys/kernel/security
-	// cp: false, because it requires the initial mount namespace to inspect file owners
```
Comment (AkihiroSuda) [inline cmd/nerdctl/main_linux.go:42]: Why was this changed?

File: cmd/nerdctl/main_linux.go [hunk 6]
```diff
@@ -39,16 +39,15 @@ func appNeedsRootlessParentMain(cmd *cobra.Command, args []string) bool {
...
+	case "", "completion", "login", "logout", "apparmor":
 		return false
 	case "container":
 		if len(commands) < 3 {
 			return true
 		}
 		switch commands[2] {
 		case "cp":
-			return false
+			return true
```
Comment (AkihiroSuda) [inline cmd/nerdctl/main_linux.go:50]: `true` is returned on L53

File: cmd/nerdctl/main_linux.go (hunk 5 above)
Comment (vsiravar) [reply cmd/nerdctl/main_linux.go:42]: I modified `appNeedsRootlessParentMain` to return true for `nerdctl cp` and `nerdctl container cp`. This was necessary to access containerd.sock in rootless mode to [create containerd client](https://github.com/containerd/nerdctl/pull/2355/files#diff-a6cc11fee8c6a93b90ff0594ba0ad1550e9e95de1f7a5f544cd47b8751c82addR64). When it's running in rootless mode without this change, get `/run/containerd/containerd.sock": no such file or directory `. However this change does not preserve the uid/gid mapping while copying to container. Should we disable the copying behaviour for stopped/created containers in rootless mode and only support this feature in rootful.

Comment (AkihiroSuda) [reply cmd/nerdctl/main_linux.go:42]: > this change does not preserve the uid/gid mapping while copying to container.

This is a regression.

> Should we disable the copying behaviour for stopped/created containers in rootless mode and only support this feature in rootful.

Yes, at least for now.

Comment (vsiravar) [reply cmd/nerdctl/main_linux.go]: @AkihiroSuda I updated the PR to 

- use rootless containerd sock based on [faq.md](https://github.com/containerd/nerdctl/blob/main/docs/faq.md#containerd-socket-address) in rootless mode.
-  Disabled copying behaviour for stopped/created containers in rootless mode and only support this feature in rootful(more info in PR description). 
-  The uid/gid mapping is preserved when files are copied. 

PTAL, thanks! I noticed that the docker-compatibility check fails although all the tests for container copy pass. Not sure if this is related to this change. PTAL, thanks for the review.

File: pkg/rootlessutil/rootlessutil_linux.go [hunk 7]
```diff
@@ -67,3 +68,16 @@ func NewRootlessKitClient() (client.Client, error) {
...
+func RootlessContainredSockAddress() (string, error) {
+	stateDir, err := RootlessKitStateDir()
+	if err != nil {
+		return "", err
+	}
+	childPid, err := RootlessKitChildPid(stateDir)
+	if err != nil {
+		return "", err
+	}
+	return path.Join(fmt.Sprintf("/proc/%d/root/run/containerd/containerd.sock", childPid)), nil
```
Comment (AkihiroSuda) [inline pkg/rootlessutil/rootlessutil_linux.go:82]:
```suggestion
	return filepath.Join(fmt.Sprintf("/proc/%d/root/run/containerd/containerd.sock", childPid)), nil
```

File: pkg/containerutil/cp_linux.go [hunk 8]
```diff
@@ -177,3 +216,33 @@ func CopyFiles(ctx context.Context, container2host bool, pid int, dst, src strin
...
+	if err != nil {
+		return "", nil, err
+	}
+	tempDir, err := os.MkdirTemp("", "nerdctl-cp-")
+	if err != nil {
+		return "", nil, err
+	}
+	err = mount.All(resp, tempDir)
+	if err != nil {
+		os.RemoveAll(tempDir)
```
Comment (AkihiroSuda) [inline pkg/containerutil/cp_linux.go:236]: This is unsafe, as tempDir may contain partial mounts

File: pkg/containerutil/cp_linux.go [hunk 9]
```diff
@@ -19,29 +19,66 @@ package containerutil
...
 	}
 	logrus.Debugf("Detected tar binary %q (GNU=%v)", tarBinary, isGNUTar)
-	var srcFull, dstFull string
-	root := fmt.Sprintf("/proc/%d/root", pid)
+	var srcFull, dstFull, root string
+	var cleanup func()
+	task, err := container.Task(ctx, nil)
+	if err != nil {
+		// if the task is simply not found, we should try to mount the snapshot. any other type of error from Task() is fatal here.
+		// Rootless does not support copying out of stopped/created containers
```
Comment (AkihiroSuda) [inline pkg/containerutil/cp_linux.go:51]: Could you add the reason?

Comment (vsiravar) [reply pkg/containerutil/cp_linux.go]: Added a comment explaining. Lmk if I need to additional details.

**Review by AkihiroSuda — APPROVED** [review]
Thanks
